src/System.IO.FileSystem/src/System/IO/DosMatcher.cs

```diff
@@ -0,0 +1,314 @@
+                            }
+                            else
+                            {
+                                sb.Append('.');
+                            }
+                            break;
+                        case '?':
+                            sb.Append('>'); // DOS_QM
+                            break;
+                        default:
+                            sb.Append(c[i]);
+                            break;
+                    }
+                }
+            }
+
+            return StringBuilderCache.GetStringAndRelease(sb);
+        }
+
+        /// <summary>
+        /// Return true if the given expression matches the given name.
+        /// </summary>
+        /// <param name="expression">The expression to match with, such as "*.foo".</param>
+        /// <param name="name">The name to check against the expression.</param>
+        /// <param name="ignoreCase">True to ignore case (default).</param>
+        /// <remarks>
+        /// This is based off of System.IO.PatternMatcher used in FileSystemWatcher, which is based off
+        /// of RtlIsNameInExpression, which defines the rules for matching DOS wildcards ('*', '?', '&lt;', '&gt;', '"').
+        /// 
+        /// Like PatternMatcher, matching will not line up with Win32 behavior unless you transform the expression
+        /// using <see cref="TranslateExpression(string)"/>
+        /// </remarks>
+        public unsafe static bool MatchPattern(string expression, ReadOnlySpan<char> name, bool ignoreCase = true)
+        {
+            string debug = new string(name);
+
+            // The idea behind the algorithm is pretty simple. We keep track of all possible locations
+            // in the regular expression that are matching the name. When the name has been exhausted,
+            // if one of the locations in the expression is also just exhausted, the name is in the
+            // language defined by the regular expression.
+
+            if (string.IsNullOrEmpty(expression) || name.Length == 0)
+                return false;
+
+            if (expression[0] == '*')
+            {
+                // Just * matches everything
+                if (expression.Length == 1)
+                    return true;
+
+                if (expression.IndexOfAny(s_wildcardChars, startIndex: 1) == -1)
+                {
+                    // Handle the special case of a single  starting *, which essentially means "ends with"
+
+                    // If the name doesn't have enough characters to match the remaining expression, it can't be a match.
+                    if (name.Length < expression.Length - 1)
+                        return false;
+
+                    // See if we end with the expression (minus the *, of course)
+                    return name.EndsWithOrdinal(expression.AsReadOnlySpan().Slice(1), ignoreCase);
+                }
+            }
+
+            Span<int> temp;
```

Comment:
>Span<int> temp; [](start = 12, length = 15)

Try making this `Span<int> temp = stackalloc int[0];    

Otherwise the local is classified as returnable and later you cannot mix it with stackallocated ones

stackalloc int[0]; is basically a noop, but will have effect of marking the `temp` as not returnable.

Reply:
Then you can do 
```C#
Span<int> priorMatches = stackalloc int[16];
Span<int> currentMatches = stackalloc int[16];
```
and remove unsafe.

---
In reply to: [152644870](https://github.com/dotnet/corefx/pull/25426#discussion_r152644870) [](ancestors = 152644870)

Reply:
That works, thanks!

Reply:
@VSadov could the compiler be improved to avoid requiring this "no op" trick?

Reply:
@danmosemsft  not easily. It is hard in general to infer returnability of a variable from its use. Especially considering that it could be assigned to other locals and they can be assigned to this one. Inference may help some cases like this one, but make other cases less intuitive.  

We clearly have an issue here. A good solution has been elusive. 
See for example: https://github.com/dotnet/csharplang/issues/1130